runtime/compiler/env/j9methodServer.cpp

```diff
@@ -1651,6 +1652,102 @@ TR_ResolvedJ9JITaaSServerMethod::addValidationRecordForCachedResolvedMethod(cons
+      {
+      // Identify all bytecodes that require a resolved method
+      int32_t cpIndex = bci.next2Bytes();
+      TR_ResolvedMethodType type = TR_ResolvedMethodType::NoType;
+      TR_ResolvedMethod *resolvedMethod;
+      switch (bc)
+         {
+         case J9BCinvokevirtual:
+            {
+            type = TR_ResolvedMethodType::VirtualFromCP;
+            break;
+            }
+         case J9BCinvokestaticsplit:
+            {
+            // falling through on purpose
+            cpIndex |= J9_STATIC_SPLIT_TABLE_INDEX_FLAG;
+            }
+         case J9BCinvokestatic:
+            {
+            type = TR_ResolvedMethodType::Static;
+            break;
+            }
+         case J9BCinvokespecialsplit:
+            {
+            // falling through on purpose
+            cpIndex |= J9_SPECIAL_SPLIT_TABLE_INDEX_FLAG;
+            }
+         case J9BCinvokespecial:
+            {
+            type = TR_ResolvedMethodType::Special;
+            break;
+            }
+         default:
+            {
+            // do nothing
+            break;
+            }
+         }
+
+      if (type != TR_ResolvedMethodType::NoType &&
+          !compInfoPT->getCachedResolvedMethod(
+             compInfoPT->getResolvedMethodKey(type, (TR_OpaqueClassBlock *) _ramClass, cpIndex),
+             this,
+             &resolvedMethod))
+         {
+         methodTypes.push_back(type);
+         cpIndices.push_back(cpIndex);
+         }
+      }
+
+   int32_t numMethods = methodTypes.size();
+   // If less than 2 methods, it's cheaper to create
+   // resolved method normally, because client won't
+   // have to deal with vectors
+   if (numMethods < 2)
+      return;
+
+   // 2. Send a remote query to mirror all uncached resolved methods
+   _stream->write(JITaaS::J9ServerMessageType::ResolvedMethod_getMultipleResolvedMethods, (TR_ResolvedJ9Method *) _remoteMirror, methodTypes, cpIndices);
+   auto recv = _stream->read<std::vector<J9Method *>, std::vector<uint32_t>, std::vector<TR_ResolvedJ9JITaaSServerMethodInfo>>();
+
+   // 3. Cache all received resolved methods
+   auto ramMethods = std::get<0>(recv);
+   auto vTableOffsets = std::get<1>(recv);
+   auto methodInfos = std::get<2>(recv);
+   TR_ASSERT(numMethods == ramMethods.size(), "Number of received methods does not match the number of requested methods");
+   for (int32_t i = 0; i < numMethods; ++i)
+      {
+      TR_ResolvedMethodType type = methodTypes[i];
+      TR_ResolvedMethod *resolvedMethod;
+      if (std::get<0>(methodInfos[i]).remoteMirror &&
+         !compInfoPT->getCachedResolvedMethod(
+             compInfoPT->getResolvedMethodKey(type, (TR_OpaqueClassBlock *) _ramClass, cpIndices[i]),
+             this,
+             &resolvedMethod))
+         {
+         compInfoPT->cacheResolvedMethod(
+            compInfoPT->getResolvedMethodKey(type, (TR_OpaqueClassBlock *) _ramClass, cpIndices[i]),
```

Comment:
Can we create a local and reuse it here and above? It's not an expensive operation, but it also can't be inlined and probably can't be optimized away.

Reply:
Done.